File: Queue/queue.py
# Basically a circular dynamic array

class Queue:
    def __init__(self):
        self.capacity = 4
        self.data = [None] * self.capacity

        self.front_idx = 0
        self.rear_idx = 0
        self.size = 0

    def resize(self):
        new_capacity = self.capacity * 2
        new_data = [None] * new_capacity

        for i in range(self.size):
            new_data[i] = self.data[(self.front_idx + i) % self.capacity]

        self.data = new_data
        self.capacity = new_capacity

        self.front_idx = 0
        self.rear_idx = self.size

    def enqueue(self, value):
        if self.size == self.capacity:
            self.resize()

        self.data[self.rear_idx] = value

        self.rear_idx = (self.rear_idx + 1) % self.capacity
        self.size += 1

    def dequeue(self):
        if self.is_empty():
            raise IndexError("Queue is empty")

        value = self.data[self.front_idx]

        self.data[self.front_idx] = None

        self.front_idx = (self.front_idx + 1) % self.capacity
        self.size -= 1

        return value

    def front(self):
        if self.is_empty():
            raise IndexError("Queue is empty")

        return self.data[self.front_idx]

    def is_empty(self):
        return self.size == 0

    def get_size(self):
        return self.size

File: Queue/queue.cpp
template <typename T>
class Queue {
private:
    T* data;

    int capacity;
    int size;

    int frontIdx;
    int rearIdx;

    void resize() {
        int newCapacity = capacity * 2;

        T* newData = new T[newCapacity];

        // Copy elements in correct logical order
        for (int i = 0; i < size; i++) {
            newData[i] = data[(frontIdx + i) % capacity];
        }

        delete[] data;

        data = newData;
        capacity = newCapacity;

        frontIdx = 0;
        rearIdx = size;
    }

public:
    Queue() {
        capacity = 4;
        size = 0;

        data = new T[capacity];

        frontIdx = 0;
        rearIdx = 0;
    }

    ~Queue() {
        delete[] data;
    }

    void enqueue(const T& value) {
        if (size == capacity) {
            resize();
        }

        data[rearIdx] = value;

        rearIdx = (rearIdx + 1) % capacity;

        size++;
    }

    T dequeue() {
        if (isEmpty()) {
            throw runtime_error("Queue is empty");
        }

        T value = data[frontIdx];

        frontIdx = (frontIdx + 1) % capacity;

        size--;

        return value;
    }

    T& front() {
        if (isEmpty()) {
            throw runtime_error("Queue is empty");
        }

        return data[frontIdx];
    }

    bool isEmpty() const {
        return size == 0;
    }

    int getSize() const {
        return size;
    }
};
